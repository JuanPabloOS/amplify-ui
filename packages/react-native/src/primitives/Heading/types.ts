import { StyleProp, TextProps, TextStyle } from 'react-native';

export type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

export interface HeadingProps extends TextProps {
  headingStyle?: StyleProp<TextStyle>;
  level?: HeadingLevel;
  truncated?: boolean;
}

export type HeadingStyles = {
  text: TextStyle;
} & {
  [key in HeadingLevel]: TextStyle;
};
