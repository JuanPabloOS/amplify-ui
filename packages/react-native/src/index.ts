export * from './primitives';
export * as primitives from './primitives';

export { InAppMessageDisplay, withInAppMessaging } from './InAppMessaging';
export {
  InAppMessagingProvider,
  useInAppMessaging,
} from '@aws-amplify/ui-react-core';
