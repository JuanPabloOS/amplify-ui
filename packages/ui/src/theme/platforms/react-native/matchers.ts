import { DesignToken } from 'style-dictionary/types/DesignToken';

const isRadius = (token: DesignToken): boolean =>
  token.path[0] === 'radii' ||
  token.path[token.path.length - 1] === 'borderRadius';

const isFontSize = (token: DesignToken): boolean =>
  token.path[0] === 'fontSizes' ||
  token.path[token.path.length - 1] === 'fontSize';

const isFontWeight = (token: DesignToken): boolean =>
  token.path[0] === 'fontWeights' ||
  token.path[token.path.length - 1] === 'fontWeight';

const isSpace = (token: DesignToken): boolean =>
  token.path[0] === 'space' ||
  token.path[token.path.length - 1].startsWith('padding');

const isBorderWidth = (token: DesignToken): boolean =>
  token.path[0] === 'borderWidths' ||
  token.path[token.path.length - 1] === 'borderWidth';

const isOpacity = (token: DesignToken): boolean =>
  token.path[0] === 'opacities' ||
  token.path[token.path.length - 1] === 'opacity';

const isTime = (token: DesignToken): boolean =>
  token.path[0] === 'time' ||
  token.path[token.path.length - 1] === 'transitionDuration';

const isShadow = (token: DesignToken): boolean => token.path[0] === 'shadows';

export {
  isBorderWidth,
  isFontSize,
  isFontWeight,
  isOpacity,
  isRadius,
  isShadow,
  isSpace,
  isTime,
};
