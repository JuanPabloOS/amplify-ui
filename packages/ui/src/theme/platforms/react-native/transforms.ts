import { DesignToken, Transform } from 'style-dictionary';
import {
  isBorderWidth,
  isFontSize,
  isFontWeight,
  isOpacity,
  isRadius,
  isShadow,
  isSpace,
  isTime,
} from './matchers';

/*
TODO: * remove transitionDuration - not supported in RN
      * shadows
      * inherit
      * auto
      * padding
      * gap
      * boxShadow
      * remove any units (px, rem, em)
*/

export const ReactNativeTransforms: Record<string, Transform> = {
  RNFontWeight: {
    type: 'value',
    matcher: isFontWeight,
    transformer: (token: DesignToken) => {
      const intValue = parseInt(token.value, 10);
      if (isNaN(intValue)) {
        return token.value;
      } else {
        return intValue.toString();
      }
    },
  },
  RNFontSize: {
    type: 'value',
    matcher: isFontSize,
    transformer: (token) => {
      return Math.floor(parseFloat(token.value) * 16);
    },
  },
  RNBorderRadius: {
    type: 'value',
    matcher: isRadius,
    transformer: (token) => {
      return Math.floor(parseFloat(token.value) * 16);
    },
  },
  RNBorderWidth: {
    type: 'value',
    matcher: isBorderWidth,
    transformer: (token) => {
      return parseInt(token.value, 10);
    },
  },
  RNSpace: {
    type: 'value',
    matcher: isSpace,
    transformer: (token) => {
      // Note: we are multiplying values by 8 instead of 16 because
      // we want the mobile spacing to be smaller.
      return Math.floor(parseFloat(token.value) * 8);
    },
  },
  RNOpacity: {
    type: 'value',
    matcher: isOpacity,
    transformer: (token) => {
      return parseFloat(token.value);
    },
  },
  RNTime: {
    type: 'value',
    matcher: isTime,
    transformer: (token) => {
      return parseInt(token.value, 10);
    },
  },
  RNShadow: {
    type: 'value',
    matcher: isShadow,
    transitive: true,
    transformer: (token) => {
      const {
        offsetX = '',
        offsetY = '',
        blurRadius = '',
        color = '',
      } = token.value;
      return {
        shadowColor: color,
        shadowOffset: {
          width: parseInt(offsetX, 10),
          height: parseInt(offsetY, 10),
        },
        shadowRadius: parseInt(blurRadius, 10),
      };
    },
  },
};
