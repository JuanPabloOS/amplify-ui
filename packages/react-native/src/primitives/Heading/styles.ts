import { StyleSheet } from 'react-native';
import { HeadingStyles } from './types';

export const styles: HeadingStyles = StyleSheet.create({
  text: { alignSelf: 'center' },
});
