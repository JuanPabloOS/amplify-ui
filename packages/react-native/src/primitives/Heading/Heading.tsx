import React from 'react';
import { Text } from 'react-native';
import { styles } from './styles';
import { HeadingProps } from './types';

export default function Heading({
  children,
  headingStyle,
  level = 6,
  truncated = false,
}: HeadingProps): JSX.Element {
  return (
    <Text
      numberOfLines={truncated ? 1 : 0}
      style={[styles.text, styles[level], headingStyle]}
    >
      {children}
    </Text>
  );
}
