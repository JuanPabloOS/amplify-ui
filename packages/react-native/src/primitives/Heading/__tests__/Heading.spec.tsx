import React from 'react';
import TestRenderer from 'react-test-renderer';
import Heading from '../Heading';
import { Text, TextStyle } from 'react-native';

const title = 'Test Heading';

describe('Heading', () => {
  it('renders a level 6 Heading by default', () => {
    const defaultHeading = TestRenderer.create(<Heading>{title}</Heading>);
    const headingLevel6 = TestRenderer.create(
      <Heading level={6}>{title}</Heading>
    );

    expect(defaultHeading.toJSON()).toEqual(headingLevel6.toJSON());
  });

  it('renders six different Heading levels', () => {
    const heading = TestRenderer.create(
      <>
        <Heading level={1}>Heading level 1</Heading>
        <Heading level={2}>Heading level 2</Heading>
        <Heading level={3}>Heading level 3</Heading>
        <Heading level={4}>Heading level 4</Heading>
        <Heading level={5}>Heading level 5</Heading>
        <Heading level={6}>Heading level 6</Heading>
      </>
    );
    expect(heading.toJSON()).toMatchSnapshot();
  });

  it('truncates long Headings with an ellipsis', () => {
    const headings = TestRenderer.create(
      <>
        <Heading numberOfLines={1}>
          Really long heading that should be truncated with an ellipsis
        </Heading>
        <Heading>
          Really long heading that should NOT be truncated with an ellipsis
        </Heading>
      </>
    );

    const [truncated, notTruncated] = headings.root.findAllByType(Text);

    expect(headings.toJSON()).toMatchSnapshot();
    expect(truncated.props.numberOfLines).toBe(1);
    expect(notTruncated.props.numberOfLines).toBe(undefined);
  });

  it('applies style props', () => {
    const customStyle = { color: 'red' };

    const styledHeading = TestRenderer.create(
      <Heading style={customStyle}>{title}</Heading>
    );

    const styleProps = styledHeading.root.findByType(Text).props
      .style as TextStyle[];

    expect(styledHeading.toJSON()).toMatchSnapshot();
    expect(styledHeading.root.props.style).toBe(customStyle);
    expect(styleProps[styleProps.length - 1]).toBe(customStyle);
  });
});
