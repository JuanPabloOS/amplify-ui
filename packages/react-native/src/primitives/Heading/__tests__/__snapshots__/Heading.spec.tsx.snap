// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading applies style props 1`] = `
<Text
  numberOfLines={0}
  style={
    Array [
      Object {
        "color": "black",
      },
      Object {
        "fontSize": 16,
        "fontWeight": "900",
      },
      Object {
        "color": "red",
      },
    ]
  }
>
  Test Heading
</Text>
`;

exports[`Heading renders six different Heading levels 1`] = `
Array [
  <Text
    numberOfLines={0}
    style={
      Array [
        Object {
          "color": "black",
        },
        Object {
          "fontSize": 56,
          "fontWeight": "100",
        },
        undefined,
      ]
    }
  >
    Heading level 1
  </Text>,
  <Text
    numberOfLines={0}
    style={
      Array [
        Object {
          "color": "black",
        },
        Object {
          "fontSize": 44,
          "fontWeight": "200",
        },
        undefined,
      ]
    }
  >
    Heading level 2
  </Text>,
  <Text
    numberOfLines={0}
    style={
      Array [
        Object {
          "color": "black",
        },
        Object {
          "fontSize": 34,
          "fontWeight": "300",
        },
        undefined,
      ]
    }
  >
    Heading level 3
  </Text>,
  <Text
    numberOfLines={0}
    style={
      Array [
        Object {
          "color": "black",
        },
        Object {
          "fontSize": 26,
          "fontWeight": "500",
        },
        undefined,
      ]
    }
  >
    Heading level 4
  </Text>,
  <Text
    numberOfLines={0}
    style={
      Array [
        Object {
          "color": "black",
        },
        Object {
          "fontSize": 20,
          "fontWeight": "700",
        },
        undefined,
      ]
    }
  >
    Heading level 5
  </Text>,
  <Text
    numberOfLines={0}
    style={
      Array [
        Object {
          "color": "black",
        },
        Object {
          "fontSize": 16,
          "fontWeight": "900",
        },
        undefined,
      ]
    }
  >
    Heading level 6
  </Text>,
]
`;

exports[`Heading truncates long Headings with an ellipsis 1`] = `
Array [
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "black",
        },
        Object {
          "fontSize": 16,
          "fontWeight": "900",
        },
        undefined,
      ]
    }
  >
    Really long heading that should be truncated with an ellipsis
  </Text>,
  <Text
    numberOfLines={0}
    style={
      Array [
        Object {
          "color": "black",
        },
        Object {
          "fontSize": 16,
          "fontWeight": "900",
        },
        undefined,
      ]
    }
  >
    Really long heading that should NOT be truncated with an ellipsis
  </Text>,
]
`;
